### ISSUE TITLE
Incorrect/Extra Data included in sbom.

### ISSUE THREAD

**Issue by dlorenc** [CycloneDX/cyclonedx-gomod#20]
I tried out this tool on an open source repo I maintain called `cosign`. The repo is here: https://github.com/sigstore/cosign

I ran the tool at cosign commit `749c7e3e5d80f3fa976f31084317a556718c3e54`.
The cyclonedx-gomod version I used was `35b214b43eabfdf6b47499427b93c23ae1f903aa`.

I ran: `$ cyclonedx-gomod -json > gomod.json`

The resulting sbom is available here: https://gist.github.com/dlorenc/17c0582acb0560807e561e436cf76a1e

The tool worked,  but a quick look at the data shows some packages that are not included in my application. For example:

```json
"components": [
    {
      "bom-ref": "pkg:golang/bazil.org/fuse@v0.0.0-20180421153158-65cc252bf669",
      "type": "library",
      "name": "bazil.org/fuse",
      "version": "v0.0.0-20180421153158-65cc252bf669",
      "scope": "required",
      "hashes": [
        {
          "alg": "SHA-256",
          "content": "14d091a578aab86d5aa32a9c2165459a94d2295731d9b403dfcb9965e1ad765c"
        }
      ],
      "purl": "pkg:golang/bazil.org/fuse@v0.0.0-20180421153158-65cc252bf669"
    },
```

bazil.org/fuse does not appear in my application at all, including it's dependency tree. I do not vendor my code normally (I saw the warnings about this tool only partially supporting vendoring). But if I run `go mod vendor`, I do not see bazil.org/fuze anywhere in the vendor tree. I can build and run my code without that package, so I think it's a bug that it appears in the SBOM.

You can see a test branch with the vendor directory here: https://github.com/dlorenc/cosign/tree/vendor/vendor/github.com

Note that this is just the first component in the list and the first that I checked. There may be other things missing or incorrect.

Am I misunderstanding the fields? Is there some reason this is included?

Note: This **does** appear in my go.sum file, but not in the vendor directory and it never makes it into my compiled application, as can be shown by the vendor directory. This means it is used somehow in the dependency constraint calculation by "go mod".

Note that there are

**Comment by nscuro** [comment]
Hi @dlorenc, thank you for the detailed report. 

What you're seeing is the result of *cyclonedx-gomod* (or Go's own module tooling rather) not knowing which modules end up in the final product. 

We use `go list -m` and `go mod graph`. Both operate on the module graph, which is decoupled from the build. Go's module graph considers *all* modules, including transitive test dependencies (ref https://github.com/golang/go/issues/26955).

Regarding `bazil.org/fuse` specifically, the dependency chain appears to be:

```
github.com/sigstore/cosign@v0.0.0-20210521125913-749c7e3e5d80
|-+ github.com/sigstore/rekor@v0.1.2-0.20210519014330-b5480728bde6
  |-+ gocloud.dev@v0.23.0
    |-+ github.com/GoogleCloudPlatform/cloudsql-proxy@v1.22.0
      |-+ bazil.org/fuse@v0.0.0-20180421153158-65cc252bf669
```

[Searching](https://github.com/sigstore/rekor/search?q=gocloud.dev) `github.com/sigstore/rekor` for `gocloud.dev` showed that it's only ever used in `cmd/rekor-server`, which in turn is [not used](https://github.com/sigstore/cosign/search?q=rekor%2Fcmd%2F) by `github.com/sigstore/cosign`. 

Based on that, `gocloud.dev` and all its dependencies won't ever end up in your binary. But because it's a dependency of the `github.com/sigstore/rekor` module, it ends up in the module graph and consequently in your SBOM.

`go mod vendor` won't copy `gocloud.dev` to `vendor/`, because in contrast to `go list -m` and `go mod graph`, it "knows" what's required to *build* your module. But even then, it'll consider both of your commands (`cmd/cosgin`, `cmd/sget`), which probably don't depend on the same modules either. 

I think we could use something along the lines of `go list -deps -f '{{with .Module}}{{.Path}}@{{.Version}}{{end}}' ./cmd/cosign | sort -u` to figure out which modules are required to build `cmd/cosign`. With that knowledge, we should be able to tidy up the SBOM quite a bit. I will experiment with this a bit.

<details>
  <summary>This is what I get for the "go list" command above. Looks a bit more reasonable.</summary>

```
cloud.google.com/go@v0.81.0
github.com/asaskevich/govalidator@v0.0.0-20210307081110-f21760c49a8d
github.com/blang/semver@v3.5.1+incompatible
github.com/cavaliercoder/go-rpm@v0.0.0-20200122174316-8cb9fd9c31a8
github.com/containerd/stargz-snapshotter/estargz@v0.4.1
github.com/coreos/go-oidc/v3@v3.0.0
github.com/cyberphone/json-canonicalization@v0.0.0-20210303052042-6bc126869bf4
github.com/danieljoos/wincred@v1.1.0
github.com/docker/cli@v0.0.0-20191017083524-a8ff7f821017
github.com/docker/distribution@v2.7.1+incompatible
github.com/docker/docker@v1.4.2-0.20200319182547-c7ad2b866182
github.com/docker/docker-credential-helpers@v0.6.3
github.com/fsnotify/fsnotify@v1.4.9
github.com/ghodss/yaml@v1.0.0
github.com/go-chi/chi@v4.1.2+incompatible
github.com/golang/groupcache@v0.0.0-20210331224755-41bb18bfe9da
github.com/golang/protobuf@v1.5.2
github.com/golang/snappy@v0.0.3
github.com/google/certificate-transparency-go@v1.1.1
github.com/google/go-cmp@v0.5.5
github.com/google/go-containerregistry@v0.5.1
github.com/google/trillian@v1.3.14-0.20210413093047-5e12fb368c8f
github.com/googleapis/gax-go/v2@v2.0.5
github.com/go-openapi/analysis@v0.20.1
github.com/go-openapi/errors@v0.20.0
github.com/go-openapi/jsonpointer@v0.19.5
github.com/go-openapi/jsonreference@v0.19.5
github.com/go-openapi/loads@v0.20.2
github.com/go-openapi/runtime@v0.19.28
github.com/go-openapi/spec@v0.20.3
github.com/go-openapi/strfmt@v0.20.1
github.com/go-openapi/swag@v0.19.15
github.com/go-openapi/validate@v0.20.2
github.com/go-playground/locales@v0.13.0
github.com/go-playground/universal-translator@v0.17.0
github.com/go-playground/validator@v9.31.0+incompatible
github.com/go-stack/stack@v1.8.0
github.com/hashicorp/errwrap@v1.1.0
github.com/hashicorp/go-cleanhttp@v0.5.2
github.com/hashicorp/go-multierror@v1.1.1
github.com/hashicorp/go-retryablehttp@v0.6.8
github.com/hashicorp/go-rootcerts@v1.0.2
github.com/hashicorp/go-sockaddr@v1.0.2
github.com/hashicorp/hcl@v1.0.0
github.com/hashicorp/vault/api@v1.1.0
github.com/hashicorp/vault/sdk@v0.2.0
github.com/howeyc/gopass@v0.0.0-20190910152052-7cb4b85ec19c
github.com/inconshreveable/mousetrap@v1.0.0
github.com/jedisct1/go-minisign@v0.0.0-20210414164026-819d7e2534ac
github.com/josharian/intern@v1.0.0
github.com/leodido/go-urn@v1.2.1
github.com/magiconair/properties@v1.8.5
github.com/mailru/easyjson@v0.7.7
github.com/mitchellh/go-homedir@v1.1.0
github.com/mitchellh/mapstructure@v1.4.1
github.com/oklog/ulid@v1.3.1
github.com/opencontainers/go-digest@v1.0.0
github.com/opencontainers/image-spec@v1.0.1
github.com/opentracing/opentracing-go@v1.2.0
github.com/pelletier/go-toml@v1.9.0
github.com/peterbourgon/ff/v3@v3.0.0
github.com/pierrec/lz4@v2.6.0+incompatible
github.com/pkg/errors@v0.9.1
github.com/PuerkitoBio/purell@v1.1.1
github.com/PuerkitoBio/urlesc@v0.0.0-20170810143723-de5bf2ad4578
github.com/ryanuber/go-glob@v1.0.0
github.com/sassoftware/relic@v7.2.1+incompatible
github.com/segmentio/ksuid@v1.0.3
github.com/sigstore/cosign@
github.com/sigstore/fulcio@v0.0.0-20210405115948-e7630f533fca
github.com/sigstore/rekor@v0.1.2-0.20210519014330-b5480728bde6
github.com/sigstore/sigstore@v0.0.0-20210427115853-11e6eaab7cdc
github.com/skratchdot/open-golang@v0.0.0-20200116055534-eef842397966
github.com/spf13/afero@v1.6.0
github.com/spf13/cast@v1.3.1
github.com/spf13/cobra@v1.1.3
github.com/spf13/jwalterweatherman@v1.1.0
github.com/spf13/pflag@v1.0.5
github.com/spf13/viper@v1.7.1
github.com/subosito/gotenv@v1.2.0
github.com/theupdateframework/go-tuf@v0.0.0-20201230183259-aee6270feb55
github.com/zalando/go-keyring@v0.1.1
go.mongodb.org/mongo-driver@v1.5.1
go.opencensus.io@v0.23.0
go.uber.org/atomic@v1.7.0
go.uber.org/multierr@v1.6.0
go.uber.org/zap@v1.16.0
golang.org/x/crypto@v0.0.0-20210506145944-38f3c27a63bf
golang.org/x/net@v0.0.0-20210505214959-0714010a04ed
golang.org/x/oauth2@v0.0.0-20210427180440-81ed05c6b58c
golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c
golang.org/x/sys@v0.0.0-20210503173754-0981d6026fa6
golang.org/x/term@v0.0.0-20210422114643-f5beecf764ed
golang.org/x/text@v0.3.6
golang.org/x/time@v0.0.0-20210220033141-f8bda1e9f3ba
google.golang.org/api@v0.46.0
google.golang.org/genproto@v0.0.0-20210506142907-4a47615972c2
google.golang.org/grpc@v1.37.1
google.golang.org/protobuf@v1.26.0
gopkg.in/ini.v1@v1.62.0
gopkg.in/square/go-jose.v2@v2.5.1
gopkg.in/yaml.v2@v2.4.0
```
</details>

**Comment by nscuro** [comment]
Note to self: Use `go list -deps -json ./...` instead of `go list -json -m all`. This only reports modules (even more granular, packages) that end up in the compiled binary, thus excluding test-only dependencies as well. The inaccurate nature of `go list -m all` for use cases like this came up [here](https://github.com/sonatype-nexus-community/nancy/issues/228) as well.

`go list -deps` is, in contrast to `go list -m`, subject to build constraints (tags, `GOOS`, `GOARCH`, see https://github.com/golang/go/issues/27900#issuecomment-437926158). So for multiple build targets, multiple SBOMs may have to be generated. This has to be documented.

**Comment by dlorenc** [comment]
> So for multiple build targets, multiple SBOMs may have to be generated. This has to be documented.

I agree - this is unintuitive at first but a go module is not 1:1 with a compiled go "application". You can have many applications under one module, and the module deps are the set of all the applications.

**Comment by nscuro** [comment]
Following a discussion in Slack, we should support the generation of BOMs that either *do* or *don't* consider build constraints. Which variant is chosen will depend on what the BOMs are supposed to be used for:

* Variant 1: distributors will want a BOM that only contains modules that are actually present in the specific binary
* Variant 2: vendors / developers will want a BOM that contains all modules that could potentially be present in any binary
  * Bascially variant 1 w/o considering build constraints

This means that `go list -deps` only solves this partly, and we need another way to support variant 2.

**Comment by nscuro** [comment]
`go mod why -m` does what we need for variant 2, actually:

> Why shows a shortest path in the import graph from the main module to
each of the listed packages. If the `-m` flag is given, why treats the
arguments as a list of modules and finds a path to any package in each
of the modules.

It also doesn't consider build flags, which is exactly what we want.

A few examples with `cosign`:

```
$ go mod why -m bazil.org/fuse
# bazil.org/fuse
(main module does not need module bazil.org/fuse)
```

```
$ go mod why -m github.com/spf13/viper
# github.com/spf13/viper
github.com/sigstore/cosign/cmd/cosign/cli
github.com/sigstore/rekor/cmd/rekor-cli/app
github.com/spf13/viper
```

For test-only dependencies, the package list contains a `*.test node`:

```
go mod why -m github.com/stretchr/testify
# github.com/stretchr/testify
github.com/sigstore/cosign/cmd/cosign/cli
github.com/sigstore/cosign/cmd/cosign/cli.test
github.com/stretchr/testify/require
```

The only downside is that we can't differentiate between multiple commands, like we can with `go list -deps`. 
This is because only the shortest path is printed, not all possible paths (see https://github.com/golang/go/issues/41305). 

I'm thinking of implementing a short-term solution, where we query `go mod why -m` with all modules in our module graph to filter out unused dependencies. That would already make a huge difference for larger projects, and reduce the pain until we have an overall satisfying solution ready.

**Comment by nscuro** [comment]
https://github.com/CycloneDX/cyclonedx-gomod/issues/20#issuecomment-847274062 has been shipped with v0.8.0.

**Comment by nscuro** [comment]
For the record: I implemented a PoC that utilizes `go list -deps -json ./...` and generated a few SBOMs with it to compare with the `go mod why -m -vendor` approach introduced in v0.8.0. Just to verify that we indeed get very close in terms of accuracy.

An example with `proton-bridge@v1.8.0`, including a diff, can be found here: https://gist.github.com/nscuro/76a4056963e56d6c3d6b81b0fcd54d89

The SBOMs have been generated with the following commands:

```
$ go run main.go -module ../proton-bridge -reproducible -output proton-bridge-v1.8.0.bom.xml
$ go run main.go -module ../proton-bridge -reproducible -output proton-bridge-v1.8.0.dist.bom.xml -distribution
```

Where `-distribution` is for the approach that uses `go list -deps`. Commands were run on a Windows x64 machine.

As expected, the SBOM generated with `go mod why` contains a few (4) more modules, all of them due to build constraints (`// +build build_qt`, `// +build darwin`, ...). This also demonstrates that v0.8.1 already reliably filters out test-only dependencies.

**Comment by prabhu** [comment]
There seems to be different interpretations of what an SBoM should or shouldn't include. From my limited experience working with several large and small customers below are some of the things I learnt.

- [ ] Seting `scope: optional` is better than filtering out test and other optional dependencies completely. Even the information about how the application is getting tested is useful information to have and share
- [ ] Setting `aggregate` type correctly as per CycloneDX 1.3 specification is often sufficient. Not every tool can generate a `complete` SBoM so by transparently setting the aggregate type the consumers could make an informed choice to use multiple or alternative tools based on their desired aggregation level

**Comment by nscuro** [comment]
> Seting `scope: optional` is better than filtering out test and other optional dependencies completely. Even the information about how the application is getting tested is useful information to have and share

Agreed, although in many cases today, the tools that *consume* these SBOMs do not differentiate between `required`, `optional` and `excluded`, which oftentimes just produces a lot of unwanted noise. Which is why most (all?) CycloneDX tools make including optional components... optional.

**Comment by prabhu** [comment]
My own tool called [dep-scan](https://github.com/appThreat/dep-scan) can understand the scope. I agree most SCA tools are just basic and pointless that's why I've built two better tools - one free and one commercial.

**Comment by nscuro** [comment]
Removed the bug label, as "incorrect / extra data" isn't really true for the use-case that *cyclonedx-gomod* covers right now, which is module-level SBOMs. The enhancement of supporting application-level SBOMs has been moved to #44.

**Comment by nscuro** [comment]
This has been addressed with the new [`app` command](https://github.com/CycloneDX/cyclonedx-gomod#app) and will be released with `v1.0.0` very soon. It has been available since `v1.0.0-alpha.2` for a while.

**Comment by dlorenc** [comment]
Thanks!
